configfile: "config/config.yaml"

def get_reference():
    return config["reference"]["fasta"]

def get_reads(wildcards):
    R1 = config[wildcards.source][wildcards.phenotype]["R1"]
    R2 = config[wildcards.source][wildcards.phenotype]["R2"]
    return R1, R2

rule all:
    input:
        expand("results/varscan/{source}.varscan", source = ["parent", "bulk"])


rule bowtie2_index:
    input:
        fasta = get_reference()
    output:
        "results/bowtie2_index/reference.1.bt2"
    conda:
        "envs/bowtie2.yaml"
    shell:
        """
        bowtie2-build {input.fasta} results/bowtie2-index/reference
        """

rule bowtie2_align:
    input:
        R1 = lambda wildcards: get_reads(wildcards)[0],
        R2 = lambda wildcards: get_reads(wildcards)[1],
        index = "results/bowtie2_index/reference.1.bt2"
    output:
        bam = "results/bowtie2_align/{source}.{phenotype}.bam"
    conda:
        "envs/bowtie2.yaml"
    shell:
        """
        bowtie2 -x results/bowtie2-index/ -1 {input.R1} -2 {input.R2} --no-unal --no-discordant | samtools view -bS - > {output}
        """

rule mpileup:
    input:
        "results/bowtie2_align/{source}.resistant.bam",
        "results/bowtie2_align/{source}.susceptible.bam"
    output:
        "results/mpileup/{source}.mpileup"
    params:
        reference = get_reference()
    conda:
        "envs/samtools.yaml"
    shell:
        """
        samtools mpileup -f {params.reference} {input} > {output}
        """

rule varscan:
    input:
        "results/mpileup/{source}.mpileup"
    output:
        "results/varscan/{source}.varscan"
    conda:
        "envs/varscan.yaml"
    shell:
        """
        varscan mpileup2snp {input} --output-vcf 1 > {output}
        """